beacon/blockchain/process_proposal.go

```diff
@@ -280,10 +271,20 @@ func (s *Service) VerifyIncomingBlock(
 		return ErrUnexpectedBlockSlot
 	}
 
+	var nextBlockData *builder.RequestPayloadData
+	if s.shouldBuildOptimisticPayloads() {
+		// state copy makes sure that preFetchBuildDataForSuccess does not affect state
+		copiedState := state.Copy(ctx)
+		nextBlockData, err = s.preFetchBuildData(copiedState, consensusTime)
+		if err != nil {
+			return fmt.Errorf("failed preFetching data for rejection: %w", err)
+		}
+	}
```

Comment:
I dont get why this branch has to be here. i.e. we can just do this copy and fetching of build data in the case we really need to? if the changes to `state` (via `verifyStateRoot` below) will persist on this `copiedState`, might as well just make the copy only when its really needed, i.e. `err` of `verifyStateRoot` is non-nil.

Reply:
As soon as you start processing the block, the associated state can be modified. Even if the block is rejected, the state can be partially written already and so we can't use that state anymore. This is why we really need to fetch the data earlier.
As per the copy, unfortunately `preFetchBuildData` does modify the state as well (via `ProcessSlot` and `ProcessFork) so I need to protect the `state` used to verify the block from the `copiedState` used to load data for next block.

We could imagine to carry out `ProcessSlot` and `ProcessFork` only once on `state`, pause the state transition, load the data we care and only then going on with the actual `ProcessBlock`. However I feel stopping and resuming the state transition is more complex?